File: 18.EditDistanceAlgo.java
/* ----------------------------------------------------------------------------  */
/*   ( The Authentic JS/JAVA CodeBuff )
 ___ _                      _              _ 
 | _ ) |_  __ _ _ _ __ _ __| |_ __ ____ _ (_)
 | _ \ ' \/ _` | '_/ _` / _` \ V  V / _` || |
 |___/_||_\__,_|_| \__,_\__,_|\_/\_/\__,_|/ |
                                        |__/ 
 */
/* --------------------------------------------------------------------------   */
/*    Youtube: https://youtube.com/@code-with-Bharadwaj                        */
/*    Github : https://github.com/Manu577228                                  */
/*    Portfolio : https://manu-bharadwaj-portfolio.vercel.app/portfolio      */
/* -----------------------------------------------------------------------  */

import java.io.*;

public class EditDistance {
    
    // Function to calculate Edit Distance using Wagner-Fischer DP Algorithm
    public static int editDistance(String s1, String s2) {
        int m = s1.length(); // length of first string
        int n = s2.length(); // length of second string
        
        // Step 1: Create DP table of size (m+1) x (n+1)
        int[][] dp = new int[m + 1][n + 1];
        
        // Step 2: Fill base cases
        for (int i = 0; i <= m; i++) {
            dp[i][0] = i;  // converting s1[0..i] to empty string requires i deletions
        }
        for (int j = 0; j <= n; j++) {
            dp[0][j] = j;  // converting empty string to s2[0..j] requires j insertions
        }
        
        // Step 3: Fill DP table for all subproblems
        for (int i = 1; i <= m; i++) {
            for (int j = 1; j <= n; j++) {
                if (s1.charAt(i - 1) == s2.charAt(j - 1)) {
                    dp[i][j] = dp[i - 1][j - 1]; // chars match → copy diagonal
                } else {
                    dp[i][j] = 1 + Math.min(
                        dp[i - 1][j],          // deletion
                        Math.min(dp[i][j - 1], // insertion
                                 dp[i - 1][j - 1]) // replacement
                    );
                }
            }
        }
        
        // Final answer is bottom-right cell
        return dp[m][n];
    }
    
    public static void main(String[] args) throws IOException {
        // Example input
        String word1 = "kitten";
        String word2 = "sitting";
        
        // Compute edit distance
        int result = editDistance(word1, word2);
        
        // Print result
        System.out.println("Edit Distance between " + word1 + " and " + word2 + " is: " + result);
    }
}

File: 18.EditDistanceAlgo.py
# Edit Distance is the minimum number of operations (insert, delete, replace) 
# required to convert one string into another.
# The Wagner-Fischer algorithm solves this using Dynamic Programming in O(m × n) time.

# Explanation

# Suppose we want to transform string word1 into word2.

# Allowed operations are:

# Insert a character.

# Delete a character.

# Replace a character.

# We use a DP table, where dp[i][j] = minimum operations to convert 
# first i chars of word1 into first j chars of word2.

# Fill base cases:

# If one string is empty → need insertions/deletions.

# Transition:

# If characters match → no operation needed.

# Otherwise → 1 + min(insert, delete, replace)

def edit_distance(s1, s2):
    m, n = len(s1), len(s2)

    dp = [[0] * (n + 1) for _ in range(m + 1)]
    
    for i in range(m + 1):
        dp[i][0] = i
    for j in range(n + 1):
        dp[0][j] = j

    for i in range(1, m + 1):
        for j in range(1, n + 1):
            if s1[i - 1] == s2[j - 1]:
                dp[i][j] = dp[i - 1][j - 1]
            
            else:
                dp[i][j] = 1 + min(
                    dp[i - 1][j],
                    dp[i][j - 1],
                    dp[i - 1][j - 1]
                )

    return dp[m][n]

word1 = "kitten"
word2 = "sitting"
result = edit_distance(word1, word2)
print("Edit Distance between", word1, "and", word2, "is:", result)